autowarefoundation/autoware.universe
# fix(system_monitor): incorrect counter increment in CPU Usage monitor

Signed-off-by: v-nakayama7440-esol <v-nakayama7440@esol.co.jp>

## Description
There is a bug that the `cpu_monitor` does not generate a warning or error unless the CPU usage becomes more than the counter threshold.
The default parameters for monitoring CPU usage are as follows, and `cpu_monitor` is supposed to generate a warning after the CPU usage becomes 96% or more and it appears 2 consecutive times (monitor interval is about 1sec). As regards to error, `cpu_monitor` shall generate an error after the usage becomes 100% and it appears 2 consecutive times.
```yaml
    usage_warn: 0.96
    usage_error: 1.00
    usage_warn_count: 2
    usage_error_count: 2
```
However, the current implementation of `cpu_monitor` generates a warning/error after the appearance of 3 times, it's counting one extra. This has been introduced from this fix https://github.com/autowarefoundation/autoware.universe/pull/557.

Changed parameters as follows to make testing easy.
```yaml
    usage_error: 0.90
    usage_error_count: 2
```
And Observed an error was generated after count of 3. (This is a dump of CPU Usage)
```console
1662360194,170445574,...,b'\x00',cpu_monitor: CPU Usage,OK,se-osaka-dev32,CPU all: status,OK,CPU all: total,16.34%,...
1662360195,526444173,...,b'\x00',cpu_monitor: CPU Usage,OK,se-osaka-dev32,CPU all: status,OK,CPU all: total,100.00%,...
1662360196,862445342,...,b'\x00',cpu_monitor: CPU Usage,OK,se-osaka-dev32,CPU all: status,OK,CPU all: total,99.94%,...
1662360198,246452080,...,b'\x02',cpu_monitor: CPU Usage,very high load,se-osaka-dev32,CPU all: status,very high load,CPU all: total,99.94%,...
```

This PR is to fix counting one extra.

## Related links

https://github.com/autowarefoundation/autoware.universe/pull/557

## Tests performed

1. Change `usage_warn` and `usage_error` in `cpu_monitor.param.yaml` to make testing easy.
   ```yaml
       usage_warn: 0.40
       usage_error: 0.90
       usage_warn_count: 2
       usage_error_count: 2
    ```
1. Simulate CPU high load by using `stress` command.
1. Verify `cpu_monitor` generates an error after count of 2. `b'\x02'` indicates an error.
   ```console
   1662358072,778439042,...,b'\x00',cpu_monitor: CPU Usage,OK,se-osaka-dev32,CPU all: status,OK,CPU all: total,39.30%,...
   1662358074,154435411,...,b'\x00',cpu_monitor: CPU Usage,OK,se-osaka-dev32,CPU all: status,OK,CPU all: total,100.00%,...
   1662358075,486447870,...,b'\x02',cpu_monitor: CPU Usage,very high load,se-osaka-dev32,CPU all: status,very high load,CPU all: total,100.00%,...
   ```

1. Also verify `cpu_monitor` generates an warning after count of 2. `b'\x01'` indicates a warning.
   ```console
   1662357625,186469428,...,b'\x00',cpu_monitor: CPU Usage,OK,se-osaka-dev32,CPU all: status,OK,CPU all: total,32.37%,...
   1662357626,514459628,...,b'\x00',cpu_monitor: CPU Usage,OK,se-osaka-dev32,CPU all: status,OK,CPU all: total,54.64%,...
   1662357627,850450383,...,b'\x01',cpu_monitor: CPU Usage,high load,se-osaka-dev32,CPU all: status,high load,CPU all: total,53.82%,...
   ```

## Notes for reviewers



## Pre-review checklist for the PR author

The PR author **must** check the checkboxes below when creating the PR.


## In-review checklist for the PR reviewers

The PR reviewers **must** check the checkboxes below before approval.


## Post-review checklist for the PR author

The PR author **must** check the checkboxes below before merging.


After all checkboxes are checked, anyone who has write access can merge the PR.

[contribution guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/
[pull request guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/pull-request-guidelines/
